File: clients/csharp/examples/collections-escape-demo.cs
using System.Text.Json;
using DataStax.AstraDB.DataApi;
using DataStax.AstraDB.DataApi.Collections;
using DataStax.AstraDB.DataApi.Core;

namespace Examples;

public class Program
{
  static async Task Main()
  {
    // Get an existing collection
    var client = new DataAPIClient();
    var database = client.GetDatabase(
      "**API_ENDPOINT**",
      "**APPLICATION_TOKEN**"
    );
    var collection = database.GetCollection("**COLLECTION_NAME**");

    // Replace a document
    var filterBuilder = Builders<Document>.CollectionFilter;
    var filter = filterBuilder.And(
      filterBuilder.Eq("areas.r&&d", false),
      filterBuilder.Lt("costs.price&.usd", 300)
    );

    var replacement = new Document()
    {
      {
        "areas",
        new Document { { "r&d", false }, { "design", true } }
      },
      {
        "costs",
        new Document { { "price.usd", 100 }, { "price.cad", 90 } }
      },
    };

    var replaceOptions =
      new CollectionFindOneAndReplaceOptions<Document>()
      {
        Projection = Builders<Document>
          .Projection.Include("areas.r&&d")
          .Include("costs.price&.usd"),
        Sort = Builders<Document>
          .CollectionSort.Ascending("areas.r&&d")
          .Descending("costs.price&.usd"),
      };

    var result = await collection.FindOneAndReplaceAsync(
      filter,
      replacement,
      replaceOptions
    );

    Console.WriteLine(JsonSerializer.Serialize(result));
  }
}

// ==============  BOUNDARY BETWEEN EXAMPLE SNIPPETS  ==============

using System.Text.Json;
using DataStax.AstraDB.DataApi;
using DataStax.AstraDB.DataApi.Collections;
using DataStax.AstraDB.DataApi.Core;
using DataStax.AstraDB.DataApi.Utils;

namespace Examples;

public class Program
{
  static async Task Main()
  {
    // Get an existing collection
    var client = new DataAPIClient();
    var database = client.GetDatabase(
      "**API_ENDPOINT**",
      "**APPLICATION_TOKEN**"
    );
    var collection = database.GetCollection("**COLLECTION_NAME**");

    // Find a document
    var filterBuilder = Builders<Document>.CollectionFilter;
    var filter = filterBuilder.And(
      filterBuilder.Eq(
        FieldEscaping.EscapeFieldNames("areas", "r&d"),
        false
      ),
      filterBuilder.Lt(
        FieldEscaping.EscapeFieldNames("costs", "price.usd"),
        300
      )
    );

    var options = new CollectionFindOneAndReplaceOptions<Document>()
    {
      Sort = Builders<Document>
        .CollectionSort.Ascending(
          FieldEscaping.EscapeFieldNames("areas", "r&d")
        )
        .Descending(FieldEscaping.EscapeFieldNames("costs", "price.usd")),
      Projection = Builders<Document>
        .Projection.Include(
          FieldEscaping.EscapeFieldNames("areas", "r&d")
        )
        .Include(FieldEscaping.EscapeFieldNames("costs", "price.usd")),
    };

    var replacement = new Document()
    {
      {
        "areas",
        new Document { { "r&d", false }, { "design", true } }
      },
      {
        "costs",
        new Document { { "price.usd", 100 }, { "price.cad", 90 } }
      },
    };

    var result = await collection.FindOneAndReplaceAsync(
      filter,
      replacement,
      options
    );

    Console.WriteLine(JsonSerializer.Serialize(result));
  }
}

File: clients/java/examples/collections-escape-demo.java
import com.datastax.astra.client.DataAPIClient;
import com.datastax.astra.client.collections.Collection;
import com.datastax.astra.client.collections.definition.documents.Document;
import com.datastax.astra.client.collections.commands.options.CollectionFindOneAndReplaceOptions;
import com.datastax.astra.client.core.query.Filter;
import com.datastax.astra.client.core.query.Filters;
import com.datastax.astra.client.core.query.Projection;
import com.datastax.astra.client.core.query.Sort;
import java.util.Map;
import java.util.Optional;

public class Example {

  public static void main(String[] args) {
    // Get an existing collection
    Collection<Document> collection =
        new DataAPIClient("**APPLICATION_TOKEN**")
            .getDatabase("**API_ENDPOINT**")
            .getCollection("**COLLECTION_NAME**");

    // Find a document
    Filter filter =
        Filters.and(Filters.eq("areas.r&&d", false), Filters.lt("costs.price&.usd", 300));
    CollectionFindOneAndReplaceOptions options =
        new CollectionFindOneAndReplaceOptions()
            .sort(Sort.ascending("areas.r&&d"), Sort.descending("costs.price&.usd"))
            .projection(Projection.include("areas.r&&d", "costs.price&.usd"));
    Document newDocument =
        new Document()
            .append(
                "areas",
                Map.of(
                    "r&d", false,
                    "design", true))
            .append(
                "costs",
                Map.of(
                    "price.usd", 100,
                    "price.cad", 90));
    Optional<Document> result = collection.findOneAndReplace(filter, newDocument, options);
    System.out.println(result);
  }
}

// ==============  BOUNDARY BETWEEN EXAMPLE SNIPPETS  ==============

import com.datastax.astra.client.DataAPIClient;
import com.datastax.astra.client.collections.Collection;
import com.datastax.astra.client.collections.definition.documents.Document;
import com.datastax.astra.client.core.query.Filter;
import com.datastax.astra.client.core.query.Filters;
import com.datastax.astra.client.core.query.Sort;
import com.datastax.astra.client.core.query.Projection;
import java.util.Map;
import java.util.Optional;
import com.datastax.astra.internal.utils.EscapeUtils;
import com.datastax.astra.client.collections.commands.options.CollectionFindOneAndReplaceOptions;

public class Example {

  public static void main(String[] args) {
    // Get an existing collection
    Collection<Document> collection =
        new DataAPIClient("**APPLICATION_TOKEN**")
            .getDatabase("**API_ENDPOINT**")
            .getCollection("**COLLECTION_NAME**");

    // Find a document
    Filter filter =
        Filters.and(Filters.eq(EscapeUtils.escapeFieldNames("areas", "r&d"), false), Filters.lt(EscapeUtils.escapeFieldNames("costs", "price.usd"), 300));
    CollectionFindOneAndReplaceOptions options =
        new CollectionFindOneAndReplaceOptions()
            .sort(Sort.ascending(EscapeUtils.escapeFieldNames("areas", "r&d")), Sort.descending(EscapeUtils.escapeFieldNames("costs", "price.usd")))
            .projection(Projection.include(EscapeUtils.escapeFieldNames("areas", "r&d"), EscapeUtils.escapeFieldNames("costs", "price.usd")));
    Document newDocument =
        new Document()
            .append(
                "areas",
                Map.of(
                    "r&d", false,
                    "design", true))
            .append(
                "costs",
                Map.of(
                    "price.usd", 100,
                    "price.cad", 90));
    Optional<Document> result = collection.findOneAndReplace(filter, newDocument, options);
    System.out.println(result);
  }
}
